<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Finding optimal resolution of hierarchical hypotheses with treeclimbR • treeclimbR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Finding optimal resolution of hierarchical hypotheses with treeclimbR">
<meta property="og:description" content="treeclimbR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">treeclimbR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/treeclimbR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/csoneson/treeclimbR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Finding optimal resolution of hierarchical
hypotheses with treeclimbR</h1>
                        <h4 data-toc-skip class="author">Charlotte
Soneson and Ruizhu Huang</h4>
            
            <h4 data-toc-skip class="date">2024-03-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csoneson/treeclimbR/blob/HEAD/vignettes/treeclimbR.Rmd" class="external-link"><code>vignettes/treeclimbR.Rmd</code></a></small>
      <div class="hidden name"><code>treeclimbR.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>treeclimbR</code> is a method for analyzing hierarchical trees
of entities, such as phylogenies, at different levels of resolution. It
proposes multiple candidates (corresponding to different aggregation
levels) that capture the latent signal, and pinpoints branches or leaves
that contain features of interest, in a data-driven way. One motivation
for such a multi-level analysis is that the most highly resolved
entities (e.g., individual species in a microbial context) may not be
abundant enough to allow a potential abundance difference between
conditions to be reliably detected. Aggregating abundances on a higher
level in the tree can detect families of species that are closely
related and that all change (possibly weakly but) concordantly. At the
same time, blindly aggregating to a higher level across the whole tree
may imply losing the ability to pinpoint specific species with a strong
signal that may not be shared with their closest neighbors in the tree.
Taken together, this motivates the development of a data-dependent
aggregation approach, which is also allowed to aggregate different parts
of the tree at different levels of resolution.</p>
<p>If you are using <code>treeclimbR</code>, please cite <span class="citation">Huang et al. (2021)</span>, which also contains the
theoretical justifications and more details of the method.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p><code>treeclimbR</code> can be installed from Bioconductor using the
following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"treeclimbR"</span><span class="op">)</span></span></code></pre></div>
<p>It integrates seamlessly with the
<code>TreeSummarizedExperiment</code> class, which allows observed data,
feature and sample annotations, as well as a tree representing the
hierarchical relationship among features (or samples) to be stored in
the same object.</p>
</div>
<div class="section level3">
<h3 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h3>
<p>This vignette outlines the main functionality of
<code>treeclimbR</code>, using simulated example data provided with the
package. We start by loading the packages that will be needed in the
analyses below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TreeSummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/csoneson/treeclimbR" class="external-link">treeclimbR</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="differential-abundance-da-analysis">Differential abundance (DA) analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-da-analysis"></a>
</h2>
<p>The differential abundance (DA) workflow in <code>treeclimbR</code>
is suitable in situations where we have observed abundances (often
counts) of a set of entities in a set of samples, the entities can be
represented as leaves of a given tree, and we are interested in finding
entities (or groups of entities in the same subtree) whose abundance is
associated with some sample phenotype (e.g., different between two
conditions). For example, in <span class="citation">Huang et al.
(2021)</span> we studied differences in the abundance of microbial
species between babies born vaginally or via C-section. We also
investigated differences in miRNA abundances between groups of mice
receiving transaortic constriction or sham surgery, and cell type
abundance differences between conditions at different clustering
granularities. In all these cases, the entities are naturally
represented as leaves in a tree (a phylogenetic tree in the first case,
a tree where internal nodes represent miRNA duplexes, primary
transcripts, and clusters of miRNAs for the second, and a clustering
tree defined based on the average similarity between baseline
high-resolution cell clusters for the third).</p>
<div class="section level3">
<h3 id="load-and-visualize-example-data">Load and visualize example data<a class="anchor" aria-label="anchor" href="#load-and-visualize-example-data"></a>
</h3>
<p>In this vignette, we will work with a simulated data set with 30
samples (15 from each of two conditions) and 100 features. 18 of the
features are differentially abundant between the two conditions; this
information is stored in the <code>Signal</code> column of the object’s
<code>rowData</code>. The features represent leaves in a tree, and the
data is stored in a <code>TreeSummarizedExperiment</code> object. Below,
we first load the data and visualize the tree and the corresponding
data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Read data</span></span>
<span><span class="va">da_lse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"da_sim_100_30_18de.rds"</span>, </span>
<span>                              package <span class="op">=</span> <span class="st">"treeclimbR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">da_lse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 100 30 </span></span>
<span><span class="co">#&gt; metadata(1): parentNodeForSignal</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(100): t8 t85 ... t45 t92</span></span>
<span><span class="co">#&gt; rowData names(1): Signal</span></span>
<span><span class="co">#&gt; colnames(30): A_1 A_2 ... B_29 B_30</span></span>
<span><span class="co">#&gt; colData names(1): group</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (100 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (100 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span>
<span></span>
<span><span class="co">## Generate tree visualization where true signal leaves are colored orange</span></span>
<span><span class="co">## ...Find internal nodes in the subtrees where all leaves are differentially </span></span>
<span><span class="co">##    abundant. These will be colored orange.</span></span>
<span><span class="va">nds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/joinNode.html" class="external-link">joinNode</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, </span>
<span>                node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">$</span><span class="va">Signal</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">br</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/findDescendant.html" class="external-link">findDescendant</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, node <span class="op">=</span> <span class="va">nds</span>,</span>
<span>                            only.leaf <span class="op">=</span> <span class="cn">FALSE</span>, self.include <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/showNode.html" class="external-link">showNode</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, </span>
<span>                                       only.leaf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">node</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">br</span>, <span class="st">"yes"</span>, <span class="st">"no"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## ...Generate tree</span></span>
<span><span class="va">da_fig_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span>tr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, layout <span class="op">=</span> <span class="st">"rectangular"</span>, </span>
<span>                      branch.length <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">signal</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/ggtree/man/attacher.html" class="external-link">%&lt;+%</a></span> <span class="va">df_color</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>no <span class="op">=</span> <span class="st">"grey"</span>, yes <span class="op">=</span> <span class="st">"orange"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## ...Zoom into the subtree defined by a particular node. In this case, we </span></span>
<span><span class="co">##    know that all true signal leaves were sampled from the subtree defined </span></span>
<span><span class="co">##    by a particular node (stored in metadata(da_lse)$parentNodeForSignal).</span></span>
<span><span class="va">da_fig_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/scaleClade.html" class="external-link">scaleClade</a></span><span class="op">(</span><span class="va">da_fig_tree</span>, </span>
<span>                          node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">$</span><span class="va">parentNodeForSignal</span>, </span>
<span>                          scale <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract count matrix and scale each row to [0, 1]</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">da_lse</span>, <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">scale_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">count</span>, <span class="fl">1</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">xx</span> <span class="op">&lt;-</span> <span class="va">x</span></span>
<span>    <span class="va">rx</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">(</span><span class="va">xx</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">rx</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scale_count</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scale_count</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot tree and heatmap of scaled counts</span></span>
<span><span class="co">## ...Generate sample annotation</span></span>
<span><span class="va">vv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"_.*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vv</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scale_count</span><span class="op">)</span></span>
<span><span class="va">anno_c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="va">vv</span>, names <span class="op">=</span> <span class="va">vv</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/TreeHeatmap.html">TreeHeatmap</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, tree_fig <span class="op">=</span> <span class="va">da_fig_tree</span>, </span>
<span>            hm_data <span class="op">=</span> <span class="va">scale_count</span>, legend_title_hm <span class="op">=</span> <span class="st">"Scaled\ncount"</span>,</span>
<span>            column_split <span class="op">=</span> <span class="va">vv</span>, rel_width <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>            tree_hm_gap <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>            column_split_label <span class="op">=</span> <span class="va">anno_c</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="treeclimbR_files/figure-html/da-load-and-visualize-data-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="aggregate-counts-for-internal-nodes">Aggregate counts for internal nodes<a class="anchor" aria-label="anchor" href="#aggregate-counts-for-internal-nodes"></a>
</h3>
<p><code>treeclimbR</code> provides functionality to find an ‘optimal’
aggregation level at which to interpret hierarchically structured data.
Starting from the <code>TreeSummarizedExperiment</code> above
(containing the observed data as well as the tree for the features), the
first step is to calculate aggregated values (in this case, counts) for
all internal nodes. This is needed so that we can then run a
differential abundance analysis on leaves and nodes simultaneously. The
results from that analysis will then be used to find the optimal
aggregation level.</p>
<p>Here, we use the <code>aggTSE</code> function from the
<code>TreeSummarizedExperiment</code> package to calculate an aggregated
count for each internal node in the tree by summing the counts for all
its descendant leaves. For other applications, other aggregation methods
(e.g., averaging) may be more suitable. This can be controlled via the
<code>rowFun</code> argument.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get a list of all node IDs</span></span>
<span><span class="va">all_node</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/showNode.html" class="external-link">showNode</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span>, only.leaf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate counts for internal nodes</span></span>
<span><span class="va">da_tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/aggTSE.html" class="external-link">aggTSE</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">da_lse</span>, rowLevel <span class="op">=</span> <span class="va">all_node</span>, rowFun <span class="op">=</span> <span class="va">sum</span><span class="op">)</span></span>
<span><span class="va">da_tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 199 30 </span></span>
<span><span class="co">#&gt; metadata(1): parentNodeForSignal</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(199): alias_1 alias_2 ... alias_198 alias_199</span></span>
<span><span class="co">#&gt; rowData names(1): Signal</span></span>
<span><span class="co">#&gt; colnames(30): A_1 A_2 ... B_29 B_30</span></span>
<span><span class="co">#&gt; colData names(1): group</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (199 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (100 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>We see that the new <code>TreeSummarizedExperiment</code> now has 199
rows (representing the original leaves + the internal nodes).</p>
</div>
<div class="section level3">
<h3 id="perform-differential-analysis-for-leaves-and-nodes">Perform differential analysis for leaves and nodes<a class="anchor" aria-label="anchor" href="#perform-differential-analysis-for-leaves-and-nodes"></a>
</h3>
<p>Next, we perform differential abundance analysis for each leaf and
node, comparing the average abundance in the two conditions. Here, any
suitable function can be used (depending on the properties of the data
matrix). <code>treeclimbR</code> provides a convenience function to
perform the differential abundance analysis using <code>edgeR</code>,
which we will use here. We will ask the wrapper function to filter out
lowly abundant features (with a total count below 15).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run differential analysis</span></span>
<span><span class="va">da_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDA.html">runDA</a></span><span class="op">(</span><span class="va">da_tse</span>, assay <span class="op">=</span> <span class="st">"counts"</span>, option <span class="op">=</span> <span class="st">"glmQL"</span>, </span>
<span>                design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">da_tse</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, filter_min_count <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                filter_min_prop <span class="op">=</span> <span class="fl">0</span>, filter_min_total_count <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p>The output of <code>runDA</code> contains the <code>edgeR</code>
results, a list of the nodes that were dropped due to a low total count,
and the tree.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">da_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "edgeR_results" "nodes_drop"    "tree"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">da_res</span><span class="op">$</span><span class="va">edgeR_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "DGELRT"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "edgeR"</span></span>
<span></span>
<span><span class="co">## Nodes with too low total count</span></span>
<span><span class="va">da_res</span><span class="op">$</span><span class="va">nodes_drop</span></span>
<span><span class="co">#&gt;  [1] "alias_32"  "alias_33"  "alias_40"  "alias_44"  "alias_51"  "alias_68" </span></span>
<span><span class="co">#&gt;  [7] "alias_87"  "alias_97"  "alias_98"  "alias_134"</span></span></code></pre></div>
<p>Again, note that any differential abundance method can be used, as
long as it produces a data frame with at least columns corresponding to
the node number, the p-value, and the inferred effect size (only the
sign will be used). For the <code>runDA</code> output, we can generate
such a table with the <code><a href="../reference/nodeResult.html">nodeResult()</a></code> function (where the
<code>PValue</code> column contains the p-value, and the
<code>logFC</code> column provides information about the sign of the
inferred change):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nodeResult.html">nodeResult</a></span><span class="op">(</span><span class="va">da_res</span>, n <span class="op">=</span> <span class="cn">Inf</span>, type <span class="op">=</span> <span class="st">"DA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">da_tbl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 189   6</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">da_tbl</span><span class="op">)</span></span>
<span><span class="co">#&gt;           node      logFC   logCPM         F       PValue          FDR</span></span>
<span><span class="co">#&gt; alias_102  102 -0.6290963 18.47205 231.23013 2.541684e-23 4.803782e-21</span></span>
<span><span class="co">#&gt; alias_114  114 -0.5748841 17.74475 121.59366 1.177694e-16 1.112921e-14</span></span>
<span><span class="co">#&gt; alias_103  103 -0.7103916 17.17136 106.80756 1.830949e-15 1.153498e-13</span></span>
<span><span class="co">#&gt; alias_115  115 -0.5948226 17.40516 103.04368 3.822717e-15 1.806234e-13</span></span>
<span><span class="co">#&gt; alias_116  116 -0.6789849 16.78855  73.81487 2.204821e-12 8.334225e-11</span></span>
<span><span class="co">#&gt; alias_110  110 -0.8245494 16.18132  64.07680 2.489723e-11 7.842629e-10</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="find-candidates">Find candidates<a class="anchor" aria-label="anchor" href="#find-candidates"></a>
</h3>
<p>Next, <code>treeclimbR</code> proposes a set of aggregation
<em>candidates</em>, corresponding to a range of values for a threshold
parameter <span class="math inline">\(t\)</span> (see <span class="citation">(Huang et al. 2021)</span>). A candidate consists of a
set of nodes, representing a specific pattern of aggregation. In
general, higher values of <span class="math inline">\(t\)</span> lead to
aggregation further up in the tree (closer to the root).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get candidates</span></span>
<span><span class="va">da_cand</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCand.html">getCand</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_tse</span><span class="op">)</span>, score_data <span class="op">=</span> <span class="va">da_tbl</span>, </span>
<span>                   node_column <span class="op">=</span> <span class="st">"node"</span>, p_column <span class="op">=</span> <span class="st">"PValue"</span>,</span>
<span>                   threshold <span class="op">=</span> <span class="fl">0.05</span>, sign_column <span class="op">=</span> <span class="st">"logFC"</span>, message <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For a given <span class="math inline">\(t\)</span>-value, we can
indicate the corresponding candidate in the tree. Note that at this
point, we have not yet selected the optimal aggregation level, nor are
we making conclusions about which of the retained nodes show a
significant difference between the conditions. The visualization thus
simply illustrates which leaves would be aggregated together if this
candidate were to be chosen.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## All candidates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">da_cand</span><span class="op">$</span><span class="va">candidate_list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "0"    "0.01" "0.02" "0.03" "0.04" "0.05" "0.1"  "0.15" "0.2"  "0.25"</span></span>
<span><span class="co">#&gt; [11] "0.3"  "0.35" "0.4"  "0.45" "0.5"  "0.55" "0.6"  "0.65" "0.7"  "0.75"</span></span>
<span><span class="co">#&gt; [21] "0.8"  "0.85" "0.9"  "0.95" "1"</span></span>
<span></span>
<span><span class="co">## Nodes contained in the candidate corresponding to t = 0.03</span></span>
<span><span class="co">## This is a mix of leaves and internal nodes</span></span>
<span><span class="op">(</span><span class="va">da_cand_0.03</span> <span class="op">&lt;-</span> <span class="va">da_cand</span><span class="op">$</span><span class="va">candidate_list</span><span class="op">[[</span><span class="st">"0.03"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1]   1   2   5   6   7   8   9  10  11  12  15  16  17  18  21  22  23  24  25</span></span>
<span><span class="co">#&gt; [20]  26  27  28  29  30  31  34  35  36  37  38  39  41  42  43  45  46  47  48</span></span>
<span><span class="co">#&gt; [39]  49  50  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  69</span></span>
<span><span class="co">#&gt; [58]  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  88  89</span></span>
<span><span class="co">#&gt; [77]  90  91  92  93  94  95  96  99 100 108 122 117</span></span>
<span></span>
<span><span class="co">## Visualize candidate</span></span>
<span><span class="va">da_fig_tree</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_point2.html" class="external-link">geom_point2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">(</span><span class="va">node</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">da_cand_0.03</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                color <span class="op">=</span> <span class="st">"navy"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"t = 0.03"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"navy"</span>, size <span class="op">=</span> <span class="fl">7</span>, </span>
<span>                                    hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.08</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="treeclimbR_files/figure-html/da-plot-cand-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="select-the-optimal-candidate">Select the optimal candidate<a class="anchor" aria-label="anchor" href="#select-the-optimal-candidate"></a>
</h3>
<p>Finally, given the set of candidates extracted above,
<code>treeclimbR</code> can now extract the one providing the optimal
aggregation level <span class="citation">(Huang et al. 2021)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Evaluate candidates</span></span>
<span><span class="va">da_best</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evalCand.html">evalCand</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_tse</span><span class="op">)</span>, levels <span class="op">=</span> <span class="va">da_cand</span><span class="op">$</span><span class="va">candidate_list</span>, </span>
<span>                    score_data <span class="op">=</span> <span class="va">da_tbl</span>, node_column <span class="op">=</span> <span class="st">"node"</span>,</span>
<span>                    p_column <span class="op">=</span> <span class="st">"PValue"</span>, sign_column <span class="op">=</span> <span class="st">"logFC"</span><span class="op">)</span></span></code></pre></div>
<p>We can get a summary of all the candidates, as well as an indication
of which <code>treeclimbR</code> considers the optimal one, using the
<code>infoCand</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/infoCand.html">infoCand</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">da_best</span><span class="op">)</span></span>
<span><span class="co">#&gt;       t    upper_t is_valid method limit_rej level_name  best rej_leaf rej_node</span></span>
<span><span class="co">#&gt; 1  0.00 0.02142857     TRUE     BH      0.05          0 FALSE       17       17</span></span>
<span><span class="co">#&gt; 2  0.01 0.02142857     TRUE     BH      0.05       0.01  TRUE       17       14</span></span>
<span><span class="co">#&gt; 3  0.02 0.02142857     TRUE     BH      0.05       0.02  TRUE       17       14</span></span>
<span><span class="co">#&gt; 4  0.03 0.02142857    FALSE     BH      0.05       0.03 FALSE       17       14</span></span>
<span><span class="co">#&gt; 5  0.04 0.02142857    FALSE     BH      0.05       0.04 FALSE       17       14</span></span>
<span><span class="co">#&gt; 6  0.05 0.03846154    FALSE     BH      0.05       0.05 FALSE       18       13</span></span>
<span><span class="co">#&gt; 7  0.10 0.05000000    FALSE     BH      0.05        0.1 FALSE       21       14</span></span>
<span><span class="co">#&gt; 8  0.15 0.05000000    FALSE     BH      0.05       0.15 FALSE       21       14</span></span>
<span><span class="co">#&gt; 9  0.20 0.05000000    FALSE     BH      0.05        0.2 FALSE       21       14</span></span>
<span><span class="co">#&gt; 10 0.25 0.06923077    FALSE     BH      0.05       0.25 FALSE       22       13</span></span>
<span><span class="co">#&gt; 11 0.30 0.09166667    FALSE     BH      0.05        0.3 FALSE       23       12</span></span>
<span><span class="co">#&gt; 12 0.35 0.09166667    FALSE     BH      0.05       0.35 FALSE       23       12</span></span>
<span><span class="co">#&gt; 13 0.40 0.09166667    FALSE     BH      0.05        0.4 FALSE       23       12</span></span>
<span><span class="co">#&gt; 14 0.45 0.09166667    FALSE     BH      0.05       0.45 FALSE       23       12</span></span>
<span><span class="co">#&gt; 15 0.50 0.09166667    FALSE     BH      0.05        0.5 FALSE       23       12</span></span>
<span><span class="co">#&gt; 16 0.55 0.09166667    FALSE     BH      0.05       0.55 FALSE       23       12</span></span>
<span><span class="co">#&gt; 17 0.60 0.09166667    FALSE     BH      0.05        0.6 FALSE       23       12</span></span>
<span><span class="co">#&gt; 18 0.65 0.09166667    FALSE     BH      0.05       0.65 FALSE       23       12</span></span>
<span><span class="co">#&gt; 19 0.70 0.09166667    FALSE     BH      0.05        0.7 FALSE       23       12</span></span>
<span><span class="co">#&gt; 20 0.75 0.09166667    FALSE     BH      0.05       0.75 FALSE       23       12</span></span>
<span><span class="co">#&gt; 21 0.80 0.09166667    FALSE     BH      0.05        0.8 FALSE       23       12</span></span>
<span><span class="co">#&gt; 22 0.85 0.09166667    FALSE     BH      0.05       0.85 FALSE       23       12</span></span>
<span><span class="co">#&gt; 23 0.90 0.09166667    FALSE     BH      0.05        0.9 FALSE       23       12</span></span>
<span><span class="co">#&gt; 24 0.95 0.09166667    FALSE     BH      0.05       0.95 FALSE       23       12</span></span>
<span><span class="co">#&gt; 25 1.00 0.10000000    FALSE     BH      0.05          1 FALSE       22       11</span></span></code></pre></div>
<p>We can also extract a vector of significant nodes from the optimal
candidate.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/topNodes.html">topNodes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">da_best</span>, n <span class="op">=</span> <span class="cn">Inf</span>, p_value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>These nodes can then be indicated in the tree.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_fig_tree</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_point2.html" class="external-link">geom_point2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="va">node</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">da_out</span><span class="op">$</span><span class="va">node</span><span class="op">)</span>,</span>
<span>                color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="treeclimbR_files/figure-html/da-plot-significant-1.png" width="576"></p>
<p>This visualization shows that in most cases, we correctly identify
groups of leaves changing synchronously, and we don’t combine true
signal nodes with non-changing ones.</p>
<p>In this case, since we are working with simulated data, we can
estimate the (leaf-level) false discovery rate and true positive rate
for the significant nodes. <code>treeclimbR</code> will automatically
extract the descendant leaves for each of the provided nodes, and
perform the evaluation on the leaf level.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fdr.html">fdr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_tse</span><span class="op">)</span>, truth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">$</span><span class="va">Signal</span><span class="op">]</span>, </span>
<span>    found <span class="op">=</span> <span class="va">da_out</span><span class="op">$</span><span class="va">node</span>, only.leaf <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;        fdr </span></span>
<span><span class="co">#&gt; 0.05882353</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tpr.html">tpr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">rowTree</a></span><span class="op">(</span><span class="va">da_tse</span><span class="op">)</span>, truth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">da_lse</span><span class="op">)</span><span class="op">$</span><span class="va">Signal</span><span class="op">]</span>, </span>
<span>    found <span class="op">=</span> <span class="va">da_out</span><span class="op">$</span><span class="va">node</span>, only.leaf <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;       tpr </span></span>
<span><span class="co">#&gt; 0.8888889</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="differential-state-ds-analysis">Differential state (DS) analysis<a class="anchor" aria-label="anchor" href="#differential-state-ds-analysis"></a>
</h2>
<p><code>treeclimbR</code> also provides functionality for performing so
called ‘differential state analysis’ in a hierarchical setting. An
example use case for this type of analysis is in a multi-sample,
multi-condition single-cell RNA-seq data set with multiple cell types,
where we are interested in finding genes that are differentially
expressed between conditions in either a single (high-resolution) cell
type (subpopulation) or a group of similar cell subpopulations.</p>
<div class="section level3">
<h3 id="load-and-visualize-example-data-1">Load and visualize example data<a class="anchor" aria-label="anchor" href="#load-and-visualize-example-data-1"></a>
</h3>
<p>We load a simulated example data set with 20 genes and 500 cells. The
cells are assigned to 10 initial (high-resolution) clusters or
subpopulations (labelled ‘t1’ to ‘t10’), which are further
hierarchically clustered into larger meta-clusters. The hierarchical
clustering tree for the subpopulations is provided as the column tree of
the <code>TreeSummarizedExperiment</code> object. In addition, the cells
come from four different samples (two from condition ‘A’ and two from
condition ‘B’). We are interested in finding genes that are
differentially expressed between the conditions (so called ‘state
markers’) at some level of the clustering tree.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds_tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ds_sim_20_500_8de.rds"</span>, </span>
<span>                              package <span class="op">=</span> <span class="st">"treeclimbR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ds_tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 20 500 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(1): Signal</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(3): cluster_id sample_id group</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: NULL</span></span>
<span><span class="co">#&gt; rowTree: NULL</span></span>
<span><span class="co">#&gt; colLinks: a LinkDataFrame (500 rows)</span></span>
<span><span class="co">#&gt; colTree: 1 phylo tree(s) (10 leaves)</span></span>
<span></span>
<span><span class="co">## Assignment of cells to high-resolution clusters, samples and conditions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">#&gt;    cluster_id sample_id       group</span></span>
<span><span class="co">#&gt;   &lt;character&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1          t7         4           B</span></span>
<span><span class="co">#&gt; 2          t7         4           B</span></span>
<span><span class="co">#&gt; 3          t1         4           B</span></span>
<span><span class="co">#&gt; 4          t4         1           A</span></span>
<span><span class="co">#&gt; 5          t6         1           A</span></span>
<span><span class="co">#&gt; 6          t1         1           A</span></span>
<span></span>
<span><span class="co">## Tree providing the successive aggregation of the high-resolution clusters</span></span>
<span><span class="co">## into more coarse-grained ones</span></span>
<span><span class="co">## Node numbers are indicated in blue, node labels in orange</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html" class="external-link">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">node</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"darkblue"</span>,</span>
<span>               hjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html" class="external-link">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"darkorange"</span>,</span>
<span>               hjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="treeclimbR_files/figure-html/ds-load-and-visualize-data-1.png" width="700"></p>
<p>The data set contains 8 genes that are simulated to be differentially
expressed between the two conditions. Four genes are differentially
expressed in all the clusters, two in clusters <code>t2</code>,
<code>t6</code> and <code>t7</code>, and two in clusters <code>t4</code>
and <code>t5</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">$</span><span class="va">Signal</span> <span class="op">!=</span> <span class="st">"no"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span> </span>
<span><span class="co">#&gt; DataFrame with 8 rows and 1 column</span></span>
<span><span class="co">#&gt;        Signal</span></span>
<span><span class="co">#&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1         all</span></span>
<span><span class="co">#&gt; 2  t2, t6, t7</span></span>
<span><span class="co">#&gt; 3         all</span></span>
<span><span class="co">#&gt; 4  t2, t6, t7</span></span>
<span><span class="co">#&gt; 5         all</span></span>
<span><span class="co">#&gt; 6         all</span></span>
<span><span class="co">#&gt; 7      t4, t5</span></span>
<span><span class="co">#&gt; 8      t4, t5</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-counts-for-internal-nodes-1">Aggregate counts for internal nodes<a class="anchor" aria-label="anchor" href="#aggregate-counts-for-internal-nodes-1"></a>
</h3>
<p>As for the DA analysis, the first step is to aggregate the counts for
the internal nodes. In this use case, it effectively corresponds to
generating pseudobulk samples for each original sample and each internal
node in the tree.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggDS.html">aggDS</a></span><span class="op">(</span>TSE <span class="op">=</span> <span class="va">ds_tse</span>, assay <span class="op">=</span> <span class="st">"counts"</span>, sample_id <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>               group_id <span class="op">=</span> <span class="st">"group"</span>, cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>, FUN <span class="op">=</span> <span class="va">sum</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in TreeSummarizedExperiment::convertNode(tree = tree, node =</span></span>
<span><span class="co">#&gt; as.character(cell_info$cluster_id)): Multiple nodes are found to have the same</span></span>
<span><span class="co">#&gt; label.</span></span>
<span><span class="va">ds_se</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 20 4 </span></span>
<span><span class="co">#&gt; metadata(3): experiment_info agg_pars n_cells</span></span>
<span><span class="co">#&gt; assays(19): alias_1 alias_2 ... alias_18 alias_19</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(4): 4 1 2 3</span></span>
<span><span class="co">#&gt; colData names(1): group</span></span></code></pre></div>
<p>Note how there is now one assay for each node in the tree (ten for
the leaves and nine for the internal nodes). Each of these assays
contains the aggregated counts for all genes in all cells belonging to
the corresponding node, split by sample ID.</p>
<p>This object also stores information about the number of cells
contributing to each node and sample</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">ds_se</span><span class="op">)</span><span class="op">$</span><span class="va">n_cells</span></span>
<span><span class="co">#&gt;            4   1   2   3</span></span>
<span><span class="co">#&gt; alias_1   10   8  14  12</span></span>
<span><span class="co">#&gt; alias_2   17  12  10  14</span></span>
<span><span class="co">#&gt; alias_3    9   9  18  11</span></span>
<span><span class="co">#&gt; alias_4   14  13  14   7</span></span>
<span><span class="co">#&gt; alias_5   11   9  13  14</span></span>
<span><span class="co">#&gt; alias_6   15  14  13  12</span></span>
<span><span class="co">#&gt; alias_7   16  18  16  13</span></span>
<span><span class="co">#&gt; alias_8   17  13  15   7</span></span>
<span><span class="co">#&gt; alias_9    7  12  13  13</span></span>
<span><span class="co">#&gt; alias_10  12  11  12  12</span></span>
<span><span class="co">#&gt; alias_11 128 119 138 115</span></span>
<span><span class="co">#&gt; alias_12 116 108 126 103</span></span>
<span><span class="co">#&gt; alias_13  36  29  42  37</span></span>
<span><span class="co">#&gt; alias_14  26  21  28  25</span></span>
<span><span class="co">#&gt; alias_15  80  79  84  66</span></span>
<span><span class="co">#&gt; alias_16  73  67  71  53</span></span>
<span><span class="co">#&gt; alias_17  40  36  40  33</span></span>
<span><span class="co">#&gt; alias_18  25  22  27  21</span></span>
<span><span class="co">#&gt; alias_19  33  31  31  20</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="perform-differential-analysis-for-leaves-and-nodes-1">Perform differential analysis for leaves and nodes<a class="anchor" aria-label="anchor" href="#perform-differential-analysis-for-leaves-and-nodes-1"></a>
</h3>
<p>Next, we perform the differential analysis. As for the DA analysis
above, <code>treeclimbR</code> provides a convenience function for
applying <code>edgeR</code> on the aggregated counts for each node.
However, any suitable method can be used.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDS.html">runDS</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">ds_se</span>, tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span>, option <span class="op">=</span> <span class="st">"glmQL"</span>, </span>
<span>                design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ds_se</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, filter_min_count <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                filter_min_total_count <span class="op">=</span> <span class="fl">1</span>, filter_min_prop <span class="op">=</span> <span class="fl">0</span>, min_cells <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                group_column <span class="op">=</span> <span class="st">"group"</span>, message <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 0 nodes are ignored, as they don't contain at least 5 cells in at least half of the samples.</span></span></code></pre></div>
<p>As before, the output contains the <code>edgeR</code> results for
each node, the tree, and the list of nodes that were dropped because of
a too low count.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ds_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "edgeR_results" "tree"          "nodes_drop"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ds_res</span><span class="op">$</span><span class="va">edgeR_results</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "alias_1"  "alias_2"  "alias_3"  "alias_4"  "alias_5"  "alias_6" </span></span>
<span><span class="co">#&gt;  [7] "alias_7"  "alias_8"  "alias_9"  "alias_10" "alias_11" "alias_12"</span></span>
<span><span class="co">#&gt; [13] "alias_13" "alias_14" "alias_15" "alias_16" "alias_17" "alias_18"</span></span>
<span><span class="co">#&gt; [19] "alias_19"</span></span></code></pre></div>
<p>We can create a table with the node results as well. Note that this
table contains the results from all genes (features) at all nodes.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nodeResult.html">nodeResult</a></span><span class="op">(</span><span class="va">ds_res</span>, type <span class="op">=</span> <span class="st">"DS"</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ds_tbl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 380   7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ds_tbl</span><span class="op">)</span></span>
<span><span class="co">#&gt;           logFC   logCPM        F       PValue          FDR node feature</span></span>
<span><span class="co">#&gt; 3...1  1.646616 16.31850 2758.468 1.599434e-38 6.077849e-36   11       3</span></span>
<span><span class="co">#&gt; 5...2 -1.577228 16.18364 2634.842 3.949545e-38 7.504136e-36   13       5</span></span>
<span><span class="co">#&gt; 6...3 -1.633327 16.19005 2551.386 7.446819e-38 9.432638e-36   13       6</span></span>
<span><span class="co">#&gt; 1...4  1.623166 16.28457 2237.433 9.866162e-37 9.372854e-35   11       1</span></span>
<span><span class="co">#&gt; 6...5 -1.566620 16.27571 1962.617 1.293029e-35 9.827017e-34   11       6</span></span>
<span><span class="co">#&gt; 5...6 -1.529990 16.26806 1808.583 6.414163e-35 4.062303e-33   11       5</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="find-candidates-1">Find candidates<a class="anchor" aria-label="anchor" href="#find-candidates-1"></a>
</h3>
<p>The next step in the analysis is to generate a list of candidates.
This is done separately for each gene - hence, we first split the result
table above by the <code>feature</code> column, and then run
<code><a href="../reference/getCand.html">getCand()</a></code> for each subtable.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Split result table by feature</span></span>
<span><span class="va">ds_tbl_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">ds_tbl</span>, f <span class="op">=</span> <span class="va">ds_tbl</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Find candidates for each gene separately</span></span>
<span><span class="va">ds_cand_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">ds_tbl_list</span><span class="op">)</span>, </span>
<span>                       FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>                           <span class="fu"><a href="../reference/getCand.html">getCand</a></span><span class="op">(</span></span>
<span>                               tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span>,</span>
<span>                               t <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.05</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                               score_data <span class="op">=</span> <span class="va">ds_tbl_list</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                               node_column <span class="op">=</span> <span class="st">"node"</span>, </span>
<span>                               p_column <span class="op">=</span> <span class="st">"PValue"</span>, </span>
<span>                               sign_column <span class="op">=</span> <span class="st">"logFC"</span>,</span>
<span>                               message <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">$</span><span class="va">candidate_list</span></span>
<span>                       <span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ds_cand_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ds_tbl_list</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-the-optimal-candidate-1">Select the optimal candidate<a class="anchor" aria-label="anchor" href="#select-the-optimal-candidate-1"></a>
</h3>
<p>We can then find the optimal candidate using the
<code><a href="../reference/evalCand.html">evalCand()</a></code> function, and list the top hits.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds_best</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evalCand.html">evalCand</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"multiple"</span>, </span>
<span>                    levels <span class="op">=</span> <span class="va">ds_cand_list</span>, score_data <span class="op">=</span> <span class="va">ds_tbl_list</span>, </span>
<span>                    node_column <span class="op">=</span> <span class="st">"node"</span>, </span>
<span>                    p_column <span class="op">=</span> <span class="st">"PValue"</span>, </span>
<span>                    sign_column <span class="op">=</span> <span class="st">"logFC"</span>, </span>
<span>                    feature_column <span class="op">=</span> <span class="st">"feature"</span>,</span>
<span>                    limit_rej <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                    message <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                    use_pseudo_leaf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ds_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/topNodes.html">topNodes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ds_best</span>, n <span class="op">=</span> <span class="cn">Inf</span>, p_value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> </span>
<span><span class="va">ds_out</span></span>
<span><span class="co">#&gt;             logFC   logCPM         F       PValue          FDR node feature</span></span>
<span><span class="co">#&gt; 1       1.6231662 16.28457 2237.4326 9.866162e-37 9.372854e-35   11       1</span></span>
<span><span class="co">#&gt; 13     -0.2425114 15.32238   12.2388 1.163025e-03 4.130369e-03    9      13</span></span>
<span><span class="co">#&gt; 2       1.5371994 16.20796 1704.7643 2.039174e-34 1.106980e-32   13       2</span></span>
<span><span class="co">#&gt; 3       1.6466158 16.31850 2758.4677 1.599434e-38 6.077849e-36   11       3</span></span>
<span><span class="co">#&gt; 4      -1.6122045 16.25077 1017.7976 4.563616e-30 1.238696e-28   13       4</span></span>
<span><span class="co">#&gt; 5      -1.5299895 16.26806 1808.5827 6.414163e-35 4.062303e-33   11       5</span></span>
<span><span class="co">#&gt; 6      -1.5666200 16.27571 1962.6171 1.293029e-35 9.827017e-34   11       6</span></span>
<span><span class="co">#&gt; 7...8   1.3222125 16.11680 1283.8522 3.772383e-13 2.606374e-12    5       7</span></span>
<span><span class="co">#&gt; 7...9   1.6572261 16.10320  516.3906 1.766865e-24 3.357044e-23    9       7</span></span>
<span><span class="co">#&gt; 8...10 -1.5238145 16.26492  749.1655 8.049489e-12 5.184417e-11    5       8</span></span>
<span><span class="co">#&gt; 8...11 -1.6918749 16.34385  182.3817 1.730751e-16 1.289579e-15    9       8</span></span>
<span><span class="co">#&gt;               adj.p signal.node</span></span>
<span><span class="co">#&gt; 1      7.892930e-35        TRUE</span></span>
<span><span class="co">#&gt; 13     1.691673e-02        TRUE</span></span>
<span><span class="co">#&gt; 2      6.525355e-33        TRUE</span></span>
<span><span class="co">#&gt; 3      2.559094e-36        TRUE</span></span>
<span><span class="co">#&gt; 4      1.216964e-28        TRUE</span></span>
<span><span class="co">#&gt; 5      2.565665e-33        TRUE</span></span>
<span><span class="co">#&gt; 6      6.896152e-34        TRUE</span></span>
<span><span class="co">#&gt; 7...8  6.706459e-12        TRUE</span></span>
<span><span class="co">#&gt; 7...9  4.038550e-23        TRUE</span></span>
<span><span class="co">#&gt; 8...10 1.287918e-10        TRUE</span></span>
<span><span class="co">#&gt; 8...11 3.461502e-15        TRUE</span></span></code></pre></div>
<p>As expected, we detect the eight truly differentially expressed
features (feature 1-8). Furthermore, we see that they have been
aggregated to different nodes in the tree. Features 1, 3, 5 and 6 are
aggregated to node 11, features 2 and 4 to node 13, and features 7 and 8
to nodes 5 and 9. Let’s see which leaves these internal nodes correspond
to.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/findDescendant.html" class="external-link">findDescendant</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span>, node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">11</span>, <span class="fl">13</span>, <span class="fl">5</span>, <span class="fl">9</span><span class="op">)</span>, </span>
<span>                      self.include <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      use.alias <span class="op">=</span> <span class="cn">FALSE</span>, only.leaf <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>       <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/convertNode.html" class="external-link">convertNode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html" class="external-link">colTree</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span>, node <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Node_11</span></span>
<span><span class="co">#&gt;  [1] "t2"  "t7"  "t6"  "t9"  "t4"  "t8"  "t10" "t1"  "t5"  "t3" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Node_13</span></span>
<span><span class="co">#&gt; [1] "t2" "t7" "t6"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $t4</span></span>
<span><span class="co">#&gt; [1] "t4"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $t5</span></span>
<span><span class="co">#&gt; [1] "t5"</span></span></code></pre></div>
<p>Indeed, comparing this to the ground truth provided above, we see
that each gene has been aggregated to the lowest node level that
contains all the original subpopulations where the gene was simulated to
be differentially expressed between conditions A and B:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ds_tse</span><span class="op">)</span><span class="op">$</span><span class="va">Signal</span> <span class="op">!=</span> <span class="st">"no"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span> </span>
<span><span class="co">#&gt; DataFrame with 8 rows and 1 column</span></span>
<span><span class="co">#&gt;        Signal</span></span>
<span><span class="co">#&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1         all</span></span>
<span><span class="co">#&gt; 2  t2, t6, t7</span></span>
<span><span class="co">#&gt; 3         all</span></span>
<span><span class="co">#&gt; 4  t2, t6, t7</span></span>
<span><span class="co">#&gt; 5         all</span></span>
<span><span class="co">#&gt; 6         all</span></span>
<span><span class="co">#&gt; 7      t4, t5</span></span>
<span><span class="co">#&gt; 8      t4, t5</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="additional-examples">Additional examples<a class="anchor" aria-label="anchor" href="#additional-examples"></a>
</h2>
<p>Additional examples of applying <code>treeclimbR</code> to
experimental data sets can be found on <a href="https://github.com/fionarhuang/treeclimbR_article" class="external-link">GitHub</a>.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary><b> This document was executed with the following package versions
(click to expand) </b>
</summary><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-03-16 r86144)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.7.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.5.0                   dplyr_1.1.4                    </span></span>
<span><span class="co">#&gt;  [3] ggtree_3.11.1                   treeclimbR_0.99.1              </span></span>
<span><span class="co">#&gt;  [5] TreeSummarizedExperiment_2.11.0 Biostrings_2.71.4              </span></span>
<span><span class="co">#&gt;  [7] XVector_0.43.1                  SingleCellExperiment_1.25.0    </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.33.3     Biobase_2.63.0                 </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.55.3            GenomeInfoDb_1.39.9            </span></span>
<span><span class="co">#&gt; [13] IRanges_2.37.1                  S4Vectors_0.41.4               </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.49.1             MatrixGenerics_1.15.0          </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.2.0               BiocStyle_2.31.0               </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3          jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;   [3] shape_1.4.6.1               magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [5] TH.data_1.1-2               farver_2.1.1               </span></span>
<span><span class="co">#&gt;   [7] nloptr_2.0.3                rmarkdown_2.26             </span></span>
<span><span class="co">#&gt;   [9] GlobalOptions_0.1.2         fs_1.6.3                   </span></span>
<span><span class="co">#&gt;  [11] zlibbioc_1.49.3             ragg_1.3.0                 </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt;  [15] minqa_1.2.6                 RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [17] rstatix_0.7.2               htmltools_0.5.7            </span></span>
<span><span class="co">#&gt;  [19] S4Arrays_1.3.6              broom_1.0.5                </span></span>
<span><span class="co">#&gt;  [21] gridGraphics_0.5-1          SparseArray_1.3.4          </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.9                  bslib_0.6.1                </span></span>
<span><span class="co">#&gt;  [25] desc_1.4.3                  plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [27] sandwich_3.1-0              zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [29] cachem_1.0.8                igraph_2.0.3               </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.4             iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [33] pkgconfig_2.0.3             Matrix_1.6-5               </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [37] GenomeInfoDbData_1.2.11     clue_0.3-65                </span></span>
<span><span class="co">#&gt;  [39] aplot_0.2.2                 digest_0.6.35              </span></span>
<span><span class="co">#&gt;  [41] colorspace_2.1-0            ggnewscale_0.4.10          </span></span>
<span><span class="co">#&gt;  [43] patchwork_1.2.0             textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [45] ggpubr_0.6.0                labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [47] cytolib_2.15.2              fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [49] colorRamps_2.3.4            polyclip_1.10-6            </span></span>
<span><span class="co">#&gt;  [51] abind_1.4-5                 compiler_4.4.0             </span></span>
<span><span class="co">#&gt;  [53] withr_3.0.0                 doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [55] ConsensusClusterPlus_1.67.0 backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [57] BiocParallel_1.37.1         viridis_0.6.5              </span></span>
<span><span class="co">#&gt;  [59] carData_3.0-5               highr_0.10                 </span></span>
<span><span class="co">#&gt;  [61] ggforce_0.4.2               ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [63] MASS_7.3-60.2               DelayedArray_0.29.9        </span></span>
<span><span class="co">#&gt;  [65] rjson_0.2.21                FlowSOM_2.11.0             </span></span>
<span><span class="co">#&gt;  [67] diffcyt_1.23.0              tools_4.4.0                </span></span>
<span><span class="co">#&gt;  [69] ape_5.7-1                   glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [71] nlme_3.1-164                grid_4.4.0                 </span></span>
<span><span class="co">#&gt;  [73] Rtsne_0.17                  reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [75] cluster_2.1.6               generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [77] gtable_0.3.4                tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [79] car_3.1-2                   utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [81] stringr_1.5.1               foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [83] pillar_1.9.0                yulab.utils_0.1.4          </span></span>
<span><span class="co">#&gt;  [85] limma_3.59.6                circlize_0.4.16            </span></span>
<span><span class="co">#&gt;  [87] splines_4.4.0               flowCore_2.15.3            </span></span>
<span><span class="co">#&gt;  [89] tweenr_2.0.3                treeio_1.27.0              </span></span>
<span><span class="co">#&gt;  [91] lattice_0.22-5              survival_3.5-8             </span></span>
<span><span class="co">#&gt;  [93] dirmult_0.1.3-5             RProtoBufLib_2.15.1        </span></span>
<span><span class="co">#&gt;  [95] tidyselect_1.2.1            ComplexHeatmap_2.19.0      </span></span>
<span><span class="co">#&gt;  [97] locfit_1.5-9.9              knitr_1.45                 </span></span>
<span><span class="co">#&gt;  [99] gridExtra_2.3               bookdown_0.38              </span></span>
<span><span class="co">#&gt; [101] edgeR_4.1.18                xfun_0.42                  </span></span>
<span><span class="co">#&gt; [103] statmod_1.5.0               stringi_1.8.3              </span></span>
<span><span class="co">#&gt; [105] ggfun_0.1.4                 lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [107] yaml_2.3.8                  boot_1.3-30                </span></span>
<span><span class="co">#&gt; [109] evaluate_0.23               codetools_0.2-19           </span></span>
<span><span class="co">#&gt; [111] tibble_3.2.1                BiocManager_1.30.22        </span></span>
<span><span class="co">#&gt; [113] ggplotify_0.1.2             cli_3.6.2                  </span></span>
<span><span class="co">#&gt; [115] systemfonts_1.0.6           munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [117] jquerylib_0.1.4             Rcpp_1.0.12                </span></span>
<span><span class="co">#&gt; [119] png_0.1-8                   XML_3.99-0.16.1            </span></span>
<span><span class="co">#&gt; [121] parallel_4.4.0              pkgdown_2.0.7.9000         </span></span>
<span><span class="co">#&gt; [123] bitops_1.0-7                lme4_1.1-35.1              </span></span>
<span><span class="co">#&gt; [125] viridisLite_0.4.2           mvtnorm_1.2-4              </span></span>
<span><span class="co">#&gt; [127] tidytree_0.4.6              scales_1.3.0               </span></span>
<span><span class="co">#&gt; [129] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">#&gt; [131] GetoptLong_1.0.5            rlang_1.1.3                </span></span>
<span><span class="co">#&gt; [133] multcomp_1.4-25</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-huang2021" class="csl-entry">
Huang, Ruizhu, Charlotte Soneson, Pierre-Luc Germain, Thomas S B
Schmidt, Christian Von Mering, and Mark D Robinson. 2021.
<span>“treeclimbR Pinpoints the Data-Dependent Resolution of
Hierarchical Hypotheses.”</span> <em>Genome Biology</em> 22 (1): 157.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruizhu Huang, Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
